File: 04-verbose-hello-world-srv/00-verbose-hello-world-srv-common/src/main/java/com/github/jinahya/hello/miscellaneous/rfc863/Rfc863Tcp1Client.java
package com.github.jinahya.hello.miscellaneous.rfc863;

/*-
 * #%L
 * verbose-hello-world-srv-common
 * %%
 * Copyright (C) 2018 - 2023 Jinahya, Inc.
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import lombok.extern.slf4j.Slf4j;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.security.MessageDigest;
import java.util.HexFormat;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

@Slf4j
public class Rfc863Tcp1Client {

    private static final InetAddress HOST = Rfc863Tcp1Server.HOST;

    private static final int PORT = Rfc863Tcp1Server.PORT;

    static final int LENGTH = Rfc863Tcp1Server.LENGTH << 1;

    private static final String ALGORITHM = Rfc863Tcp1Server.ALGORITHM;

    public static void main(String... args) throws Exception {
        try (var client = new Socket()) {
            client.setReuseAddress(true);
            var bind = true;
            if (bind) {
                client.bind(new InetSocketAddress(HOST, 0));
                log.debug("[C] bound to {}", client.getLocalSocketAddress());
            }
            client.connect(new InetSocketAddress(HOST, PORT), (int) TimeUnit.SECONDS.toMillis(8L));
            log.debug("[C] connected to {}, through {}", client.getRemoteSocketAddress(),
                      client.getLocalSocketAddress());
            var bytes = ThreadLocalRandom.current().nextInt(1048576);
            log.debug("[C] sending {} byte(s)...", bytes);
            var digest = MessageDigest.getInstance(ALGORITHM);
            for (var buffer = new byte[LENGTH]; bytes > 0; ) {
                ThreadLocalRandom.current().nextBytes(buffer);
                var length = Math.min(buffer.length, bytes);
                client.getOutputStream().write(buffer, 0, length);
                log.trace("[C] - written: {}", length);
                bytes -= length;
                digest.update(buffer, 0, length);
            }
            client.getOutputStream().flush();
            client.shutdownOutput();
            log.debug("[S] digest: {}", HexFormat.of().formatHex(digest.digest()));
        }
    }

    private Rfc863Tcp1Client() {
        throw new AssertionError("instantiation is not allowed");
    }
}

File: 04-verbose-hello-world-srv/00-verbose-hello-world-srv-basic/src/main/zava/com/github/jinahya/hello/misc/c01rfc863/Rfc863Tcp1Client.js
/*-
 * #%L
 * verbose-hello-world-srv-common
 * %%
 * Copyright (C) 2018 - 2023 Jinahya, Inc.
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
// #!/usr/bin/node

'use strict';

const assert = require('node:assert');
const crypto = require('node:crypto');
const net = require('node:net');

function main() {
    const client = new net.Socket();
    client.on('connect', function () {
        console.log('connected to %s:%d/%s', client.remoteAddress, client.remotePort, client.remoteFamily);
        const hash = crypto.createHash('sha1');
        let bytes = Math.floor(Math.random() * 65536);
        console.log('sending %d byte(s)', bytes);
        const length = Math.floor(Math.random() * 8192) + 1;
        console.log('array.length: %d', length);
        let array = new Uint8Array(length);
        while (bytes > 0) {
            crypto.getRandomValues(array);
            if (array.length > bytes) {
                array = array.slice(0, bytes);
            }
            if (!client.write(array)) {
                client.once('drain');
            }
            hash.update(array);
            bytes -= array.length;
        }
        client.end();
        client.destroy();
        console.log('digest: %s', hash.digest('hex'));
    });
    client.on('error', function (error) {
        console.log('error: %s', error);
    });
    client.on('end', function () {
    });
    client.on('close', function () {
    });
    const ipv4 = true;
    const host = ipv4 ? '127.0.0.1' : '::1';
    const port = 50009;
    if (ipv4) assert(net.isIPv4(host)); else assert(net.isIPv6(host));
    client.connect({
        host: host,
        port: port
    });
}

if (require.main === module) {
    main();
}
